lib/Parse/Lexer.cpp

```diff
@@ -1013,34 +1020,47 @@ void Lexer::lexHexNumber() {
 void Lexer::lexNumber() {
   const char *TokStart = CurPtr-1;
   assert((isDigit(*TokStart) || *TokStart == '.') && "Unexpected start");
-  
-  auto expected_digit = [&](const char *loc, Diag<> msg) {
-    diagnose(loc, msg);
+
+  auto expected_digit = [&]() {
     while (advanceIfValidContinuationOfIdentifier(CurPtr, BufferEnd));
     return formToken(tok::unknown, TokStart);
   };
 
+  auto expected_int_digit = [&](const char *loc, unsigned kind) {
+    diagnose(loc, diag::lex_expected_digit_in_int_literal, kind);
+    return expected_digit();
+  };
+
   if (*TokStart == '0' && *CurPtr == 'x')
     return lexHexNumber();
   
   if (*TokStart == '0' && *CurPtr == 'o') {
     // 0o[0-7][0-7_]*
     ++CurPtr;
     if (*CurPtr < '0' || *CurPtr > '7')
-      return expected_digit(CurPtr, diag::lex_expected_digit_in_int_literal);
-      
+      return expected_int_digit(CurPtr, 1 /*Octal*/);
+
     while ((*CurPtr >= '0' && *CurPtr <= '7') || *CurPtr == '_')
       ++CurPtr;
+
+    if (isHexDigit(*CurPtr))
```

Comment:
You can't directly use `isValidIdentifierContinuationCodePoint` because it receives `uint32_t ` Unicode code point, not UTF8 code point.
`advanceIfValidContinuationOfIdentifier` decodes UTF8 sequence to Unicode code point.

For example, we want to tokenize `12号`  as single `tok::unkown`, no?

If you don't want to advance `CurPtr`, you could use a temporary pointer for this:

```cpp
auto tmpPtr  = CurPtr;
if (advanceIfValidContinuationOfIdentifier(tmpPtr, BufferEnd)) { ...
```

However, in most cases, you `should` consume it because `expected_digit` consumes them anyway.